File: 5_map/哈希与双向链表/432. 全 O(1) 的数据结构.ts
// 双向链表 + map
// 链表找最大最小
// key到count count到Node 的映射

// node存当前值为count的所有key
class ListNode {
  pre!: ListNode
  next!: ListNode
  count: number
  // 记录该cnt(计数)下key包括哪些
  keySet: Set<string>

  constructor(count: number = 1) {
    this.count = count
    this.keySet = new Set()
  }
}

class AllOne {
  // 记录头尾 便于求最小值最大值
  private head: ListNode
  private tail: ListNode
  private keyToCount: Map<string, number>
  private countToNode: Map<number, ListNode>

  constructor() {
    this.head = new ListNode(-Infinity)
    this.tail = new ListNode(Infinity)
    this.head.next = this.tail
    this.tail.pre = this.head
    // key对应的值
    this.keyToCount = new Map()
    // 值对应的节点
    this.countToNode = new Map()
  }

  inc(key: string): void {
    if (this.keyToCount.has(key)) {
      this.changeKey(key, 1)
    } else {
      this.keyToCount.set(key, 1)
      // 说明没有计数为1的节点,在self.head后面加入
      if (this.head.next.count !== 1) this.appendNode(new ListNode(1), this.head)

      this.head.next.keySet.add(key)
      this.countToNode.set(1, this.head.next)
    }
  }

  dec(key: string): void {
    if (this.keyToCount.has(key)) {
      const count = this.keyToCount.get(key)!
      if (count === 1) {
        this.keyToCount.delete(key)
        this.removeKeyInNode(this.countToNode.get(count)!, key)
      } else {
        this.changeKey(key, -1)
      }
    }
  }

  getMaxKey(): string {
    console.dir(this.head, { depth: null })
    return this.tail.pre === this.head ? '' : this.tail.pre.keySet.keys().next().value
  }

  getMinKey(): string {
    return this.head.next === this.tail ? '' : this.head.next.keySet.keys().next().value
  }

  // key加1或者减1
  private changeKey(key: string, offset: 1 | -1) {
    const count = this.keyToCount.get(key)!
    const node = this.countToNode.get(count)!
    const newCount = count + offset
    this.keyToCount.set(key, newCount)

    let newNode: ListNode
    if (this.countToNode.has(newCount)) {
      newNode = this.countToNode.get(newCount)!
    } else {
      newNode = new ListNode(newCount)
      this.countToNode.set(newCount, newNode)
      this.appendNode(newNode, offset === 1 ? node : node.pre)
    }
    newNode.keySet.add(key)
    this.removeKeyInNode(node, key)
  }

  // 在链表删除node
  private removeNode(node: ListNode) {
    const preNode = node.pre
    const nextNode = node.next
    preNode.next = nextNode
    nextNode.pre = preNode
  }

  // 在node删除key
  private removeKeyInNode(node: ListNode, key: string) {
    node.keySet.delete(key)
    if (node.keySet.size === 0) {
      this.removeNode(node)
      this.countToNode.delete(node.count)
    }
  }

  // 在preNode后面加入node
  private appendNode(node: ListNode, preNode: ListNode) {
    const next = preNode.next
    preNode.next = node
    node.pre = preNode
    node.next = next
    next.pre = node
  }
}

const allOne = new AllOne()

allOne.inc('hello')
allOne.inc('hello')
console.log(allOne.getMaxKey())
console.log(allOne.getMinKey())
allOne.inc('leet')
console.log(allOne.getMaxKey())
console.log(allOne.getMinKey())

File: 5_map/哈希与双向链表/go/432. 全 O(1) 的数据结构.go
// 432. 全 O(1) 的数据结构
// https://leetcode.cn/problems/all-oone-data-structure/description/?envType=problem-list-v2&envId=design
// !请你设计一个用于存储字符串计数的数据结构，并能够返回计数最小和最大的字符串。
// 实现 AllOne 类：
// AllOne() 初始化数据结构的对象。
// inc(String key) 字符串 key 的计数增加 1 。如果数据结构中尚不存在 key ，那么插入计数为 1 的 key 。
// dec(String key) 字符串 key 的计数减少 1 。如果 key 的计数在减少后为 0 ，那么需要将这个 key 从数据结构中删除。测试用例保证：在减少计数前，key 存在于数据结构中。
// getMaxKey() 返回任意一个计数最大的字符串。如果没有元素存在，返回一个空字符串 "" 。
// getMinKey() 返回任意一个计数最小的字符串。如果没有元素存在，返回一个空字符串 "" 。
// 注意：每个函数都应当满足 O(1) 平均时间复杂度。
//
// !链表的每个节点维护count.

package main

import "container/list"

type bucket struct {
	keys  map[string]struct{}
	count int
}

type AllOne struct {
	*list.List
	keyToNode map[string]*list.Element
}

func Constructor() AllOne {
	return AllOne{List: list.New(), keyToNode: make(map[string]*list.Element)}
}

func (l *AllOne) Inc(key string) {
	if curNode := l.keyToNode[key]; curNode != nil {
		curBucket := curNode.Value.(*bucket)
		if nextNode := curNode.Next(); nextNode == nil || nextNode.Value.(*bucket).count > curBucket.count+1 {
			l.keyToNode[key] = l.InsertAfter(&bucket{keys: map[string]struct{}{key: {}}, count: curBucket.count + 1}, curNode)
		} else {
			nextNode.Value.(*bucket).keys[key] = struct{}{}
			l.keyToNode[key] = nextNode
		}
		delete(curBucket.keys, key)
		if len(curBucket.keys) == 0 {
			l.Remove(curNode)
		}
	} else {
		if l.Front() == nil || l.Front().Value.(*bucket).count > 1 {
			l.keyToNode[key] = l.PushFront(&bucket{keys: map[string]struct{}{key: {}}, count: 1})
		} else {
			l.Front().Value.(*bucket).keys[key] = struct{}{}
			l.keyToNode[key] = l.Front()
		}
	}
}

func (l *AllOne) Dec(key string) {
	curNode := l.keyToNode[key]
	curBucket := curNode.Value.(*bucket)
	if curBucket.count > 1 {
		if pre := curNode.Prev(); pre == nil || pre.Value.(*bucket).count < curBucket.count-1 {
			l.keyToNode[key] = l.InsertBefore(&bucket{keys: map[string]struct{}{key: {}}, count: curBucket.count - 1}, curNode)
		} else {
			pre.Value.(*bucket).keys[key] = struct{}{}
			l.keyToNode[key] = pre
		}
	} else {
		delete(l.keyToNode, key)
	}
	delete(curBucket.keys, key)
	if len(curBucket.keys) == 0 {
		l.Remove(curNode)
	}
}

func (l *AllOne) GetMaxKey() string {
	if b := l.Back(); b != nil {
		for k := range b.Value.(*bucket).keys {
			return k
		}
	}
	return ""
}

func (l *AllOne) GetMinKey() string {
	if f := l.Front(); f != nil {
		for k := range f.Value.(*bucket).keys {
			return k
		}
	}
	return ""
}
